src/SFA.DAS.AssessorService.Application/Handlers/Certificates/GetCertificatesHistoryHandler.cs

```diff
@@ -0,0 +1,113 @@
+using SFA.DAS.AssessorService.Domain.JsonData;
+using SFA.DAS.AssessorService.ExternalApis;
+using SFA.DAS.AssessorService.ExternalApis.AssessmentOrgs;
+
+namespace SFA.DAS.AssessorService.Application.Handlers.Certificates
+{
+    public class GetCertificatesHistoryHandler : IRequestHandler<GetCertificateHistoryRequest, PaginatedList<CertificateHistoryResponse>>
+    {
+        private readonly ICertificateRepository _certificateRepository;
+        private readonly IContactQueryRepository _contactQueryRepository;
+        private readonly IAssessmentOrgsApiClient _assessmentOrgsApiClient;
+        private readonly ILogger<GetCertificatesHistoryHandler> _logger;
+
+        public GetCertificatesHistoryHandler(ICertificateRepository certificateRepository,
+            IContactQueryRepository contactQueryRepository,
+            IAssessmentOrgsApiClient assessmentOrgsApiClient,
+            ILogger<GetCertificatesHistoryHandler> logger)
+        {
+            _certificateRepository = certificateRepository;
+            _contactQueryRepository = contactQueryRepository;
+            _assessmentOrgsApiClient = assessmentOrgsApiClient;
+            _logger = logger;
+        }
+
+        public async Task<PaginatedList<CertificateHistoryResponse>> Handle(GetCertificateHistoryRequest request, CancellationToken cancellationToken)
+        {
+            const int pageSize = 10;
+
+            var contact = await _contactQueryRepository.GetContact(request.Username);
+
+            var certificates = await _certificateRepository.GetCertificateHistory(
+                contact.OrganisationId.Value,
+                request.PageIndex ?? 1,
+                pageSize);
+
+            // Please Note:- Cannot seem to automap this with custom value/type converters
+            // so dealing with it manually for now.
+            var certificateHistoryResponses = MapCertificates(certificates);
+
+            return certificateHistoryResponses;
+        }
+
+        private PaginatedList<CertificateHistoryResponse> MapCertificates(PaginatedList<Certificate> certificates)
+        {
+            var trainingProviderName = string.Empty;
+            var certificateResponses = certificates.Items.Select(
+                q =>
+                {
+                    var certificateData = JsonConvert.DeserializeObject<CertificateData>(q.CertificateData);
+
+                    try
+                    {
+                        if (q.Organisation.EndPointAssessorUkprn.HasValue)
+                        {
+                            //var provider = _assessmentOrgsApiClient
+                            //    .GetProvider(q.ProviderUkPrn).GetAwaiter()
+                            //    .GetResult();
+                            // Temporary value for now pending deployment
+                            var provider = _assessmentOrgsApiClient
+                                .GetProvider(10000268).GetAwaiter()
+                                .GetResult();
```

Comment:
I thought the whole point of storing ProviderName on the Cert Data was so that you wouldn't have to make this call?  You can just get it from CertificateData now. 😁

Reply:
yep we said yesterday we would leave it for now until we get the scripts done - is on the list to do.
TBD:-

Reply:
After discussion with @alandburns, this is what this needs to do:

- Try and get the ProviderName from the CertificateData
- If it's null, then call the api to get it and update the CertificateData with the returned ProviderName
- It it's not null, then just use it.

Reply:
done